src/globus_cli/termio/field.py

```diff
@@ -0,0 +1,110 @@
+from __future__ import annotations
+
+import abc
+import datetime
+import enum
+import typing as t
+
+
+class FieldFormatter(abc.ABC):
+    @abc.abstractmethod
+    def format(self, value: t.Any) -> str | None:
+        ...
+
+
+class _StrFieldFormatter(FieldFormatter):
+    def format(self, value: t.Any) -> str | None:
+        return str(value)
+
+
+class _DateFieldFormatter(FieldFormatter):
+    def format(self, value: t.Any) -> str | None:
+        if not value:
+            return None
+        # let this raise ValueError
+        date = datetime.datetime.fromisoformat(value)
+        if date.tzinfo is None:
+            return date.strftime("%Y-%m-%d %H:%M:%S")
+        return date.astimezone().strftime("%Y-%m-%d %H:%M:%S")
+
+
+class _BoolFieldFormatter(FieldFormatter):
+    def format(self, value: t.Any) -> str | None:
+        return str(bool(value))
+
+
+def _key_to_keyfunc(k):
+    """
+    We allow for 'keys' which are functions that map columns onto value
+    types -- they may do formatting or inspect multiple values on the
+    object. In order to support this, wrap string keys in a simple function
+    that does the natural lookup operation, but return any functions we
+    receive as they are.
```

Comment:
My gut reaction is that non-statically knowable api response types is something of a poor api design.

I haven't read through all the usages though, so totally possible I'm missing something here.

If the goal here is to take a column & transform it into something else, I'd vie for doing that in an explicit "transformer" Field parameter rather than overloading the "key" parameter.

Reply:
I agree with this, especially in light of our discussion earlier and the headway we made by introducing `FieldFormatter.parse -> FieldFormatter.render` as a paradigm for this.

That said, even if I agree 100% that API responses should be predictable, it doesn't stop the fact that there may be unusual APIs in play today. I think we need to be able to handle them. In particular, I think there are cases in which whole objects may be present or `null`, and we need to "intelligently" parse them. The default behavior we decided on for handling `null` values covers this.

I'm working on removing all non-string keys, and none of them have been particularly difficult. It's just taking time to find and update all of the various cases.